lib/push.py

```diff
@@ -0,0 +1,146 @@
+
+
+## Classes ##
+
+class Image_Pusher:
+
+   # Note; We use functions to create the blank config and manifest to to
+   # avoid copy/deepcopy complexity from just copying a default dict.
+
+   __slots__ = ("config",    # sequence of bytes
+                "dst_ref",   # destination of upload
+                "image",     # Image object we are uploading
+                "layers",    # list of (digest, .tar.gz path), lowest first
+                "manifest")  # sequence of bytes
+
+   def __init__(self, image, dst_ref):
+      self.config = None
+      self.dst_ref = dst_ref
+      self.image = image
+      self.layers = None
+      self.manifest = None
+
+   @classmethod
+   def config_new(class_):
+      "Return an empty config, ready to be filled in."
+      # FIXME: URL of relevant docs?
+      # FIXME: tidy blank/empty fields?
+      return { "architecture": platform.machine(),
+               "charliecloud_version": version.VERSION,
+               "comment": "pushed with Charliecloud",
+               "config": {},
+               "container_config": {},
+               "created": ch.now_utc_iso8601(),
+               "history": [],
+               "os": "linux",
+               "rootfs": { "diff_ids": [], "type": "layers" },
+               "weirdal": "yankovic" }
+
+   @classmethod
+   def manifest_new(class_):
+      "Return an empty manifest, ready to be filled in."
+      return { "schemaVersion": 2,
+               "mediaType": ch.TYPE_MANIFEST,
+               "config": { "mediaType": ch.TYPE_CONFIG,
+                           "size": None,
+                           "digest": None },
+               "layers": [],
+               "weirdal": "yankovic" }
+
+   def cleanup(self):
+      ch.INFO("cleaning up")
+      # Delete the tarballs since we can't yet cache them.
+      for (_, tar_c) in self.layers:
+         ch.DEBUG("deleting tarball: %s" % tar_c)
+         ch.unlink(tar_c)
+
+   def prepare(self):
+      """Prepare self.image for pushing to self.dst_ref. Return tuple: (list
+         of gzipped layer tarball paths, config as a sequence of bytes,
+         manifest as a sequence of bytes).
+
+         There is not currently any support for re-using any previously
+         prepared files already in the upload cache, because we don't yet have
+         a way to know if these have changed until they are already build."""
+      ch.mkdirs(ch.storage.upload_cache)
+      tars_uc = self.image.tarballs_write(ch.storage.upload_cache)
+      tars_c = list()
+      config = self.config_new()
+      manifest = self.manifest_new()
+      # Prepare layers.
+      for (i, tar_uc) in enumerate(tars_uc, start=1):
+         ch.INFO("layer %d/%d: preparing" % (i, len(tars_uc)))
+         path_uc = ch.storage.upload_cache // tar_uc
+         hash_uc = ch.file_hash(path_uc)
+         config["rootfs"]["diff_ids"].append("sha256:" + hash_uc)
+         #size_uc = ch.file_size(path_uc)
+         path_c = ch.file_gzip(path_uc, ["-9", "--no-name"])
```

Comment:
Is creation time relevant here as well? One issue I had with compression was the metadata bits; compressing the same uncompressed tarball resulted in different hashes. Is `--no-name` sufficient?

Reply:
Seems to do the trick:

```
$ echo foo > foo.txt
$ echo foo > bar.txt
$ ls -l *.txt
-rw-r----- 1 reidpr reidpr 4 Dec 17 16:58 bar.txt
-rw-r----- 1 reidpr reidpr 4 Dec 17 16:57 foo.txt
$ gzip -9 --no-name *.txt
$ sha256sum *.txt.gz
9a904ea72399535277c74ccf752b4c815b20100e16f9032078b73db7b86bda79  bar.txt.gz
9a904ea72399535277c74ccf752b4c815b20100e16f9032078b73db7b86bda79  foo.txt.gz
```